### DISCUSSION TITLE
On the issue of dependency package deletion

### DISCUSSION THREAD

**Discussion by lgd8981289** [pnpm/pnpm#4141] [Q&A]
I read the document of "The Original Intention of the Project":

> When using npm or Yarn, if you have 100 projects that use a dependency, 100 copies of the dependency will be saved on the hard drive. For pnpm, dependencies will be stored in a content-addressable repository, so:
> 
> If you use different versions of a dependency, only the files with differences will be added to the repository. For example, if a package has 100 files, and its new version only changes one of the files. Then pnpm update will only add 1 new file to the storage center, and will not copy the contents of the new version package just because of the change of only one file.
> All files will be stored in a certain location on the hard drive. When the package is installed, the files in the package will be hard-linked to this location without taking up additional disk space. This allows you to share the same version of dependencies across projects.
> Therefore, you save a lot of space on the disk, which is proportional to the number of projects and dependencies, and the installation speed is much faster!
> 
> Create a non-flattened node_modules folder​
> When using npm to install dependencies, all dependencies will be promoted to the root directory of the module. Therefore, the project can access dependencies that have not been added to the current project.
> 

Does this mean that once I delete a project, the shared dependencies in the project (dependency packages stored in a certain location on the hard drive) will not be deleted?
Will this cause many dependent packages that have not been referenced by any projects in my computer in the future?

---------------------------------------------

我阅读了《项目初衷》的文档：

> 当使用 npm 或 Yarn 时，如果你有100个项目使用了某个依赖，就会有100份该依赖的副本保存在硬盘上。 对于 pnpm ，依赖项将存储在一个内容可寻址的仓库中，因此：
> 
> 如果你用到了某依赖项的不同版本，那么只会将有差异的文件添加到仓库。 例如，如果某个包有100个文件，而它的新版本只改变了其中1个文件。那么 pnpm update 时只会向存储中心额外添加1个新文件，而不会因为仅仅一个文件的改变复制整新版本包的内容。
> 所有文件都会存储在硬盘上的某一位置。 当软件包被被安装时，包里的文件会硬链接到这一位置，而不会占用额外的磁盘空间。 这允许你跨项目地共享同一版本的依赖。
> 因此，您在磁盘上节省了大量空间，这与项目和依赖项的数量成正比，并且安装速度要快得多！
> 
> 创建非扁平化的 node_modules 文件夹​
> 当使用 npm 安装依赖时，所有的依赖都会被提升到模块的根目录。 因此，项目可以访问到未被添加进当前项目的依赖。

这是不是意味着一旦我删除了某个项目之后，该项目中的共享依赖（存储在硬盘上的某一位置的依赖包）并不会被删除？
这会不会导致着将来我的电脑中会出现很多没有被任何项目引用过的依赖包？

**Comment by zkochan** [comment]
Yes.

You can run `pnpm store prune` to remove the unused files from the global content-addressable store.
